From 343d108e66be73512bf7fc2ece335ea199829c7f Mon Sep 17 00:00:00 2001
From: MengLi <meng.li@windriver.com>
Date: Thu, 19 Jul 2018 17:43:12 +0800
Subject: [PATCH 1340/1345] driver: ethernet: marvell: add shutdown function

Aftering running 'kexec command, the Ethernet is not able to work
normally. Because when the secondary kernel boots up, the buffer
management module of Ethernet controller is still in active status.

Therefore, add shutdown function to desctroy BM pool so that BM
returns into initial status.

Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c |   15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c b/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c
index 27d5b7c..6bbab97 100644
--- a/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c
+++ b/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c
@@ -6310,6 +6310,20 @@ static int mvpp2x_resume(struct device *dev)
 	return 0;
 }
 
+static void mvpp2x_shutdown(struct platform_device *pdev)
+{
+	struct mv_pp2x *priv = platform_get_drvdata(pdev);
+	int i = 0;
+
+	for (i = 0; i < MVPP2_BM_SWF_NUM_POOLS; i++) {
+		struct mv_pp2x_bm_pool *bm_pool = &priv->bm_pools[i];
+
+		mv_pp2x_bm_pool_destroy(&pdev->dev, priv, bm_pool);
+	}
+
+	return;
+}
+
 static const struct dev_pm_ops mv_pp2x_pm_ops = {
 	.suspend = mvpp2x_suspend,
 	.resume = mvpp2x_resume,
@@ -6321,6 +6335,7 @@ static int mvpp2x_resume(struct device *dev)
 static struct platform_driver mv_pp2x_driver = {
 	.probe = mv_pp2x_probe,
 	.remove = mv_pp2x_remove,
+	.shutdown = mvpp2x_shutdown,
 	.driver = {
 		.name = MVPP2_DRIVER_NAME,
 		.of_match_table = mv_pp2x_match_tbl,
-- 
1.7.9.5

